templates/smtp-credentials.yaml

```diff
@@ -0,0 +1,272 @@
+      ParameterType: password
+
+  SmtpUsername:
+    Type: Custom::SmtpUsername
+    Properties:
+      ServiceToken: !GetAtt SmtpLambdaFunction.Arn
+      Key: !Ref SmtpUserAccessKey
+      ParameterType: username
+
+  SmtpLambdaExecutionRole:
+    Type: AWS::IAM::Role
+    Properties:
+      RoleName: SMTPLambdaExecutionRole
+      Description: Role assumed by Lambda to generate SMTP credentials
+      AssumeRolePolicyDocument:
+        Version: 2012-10-17
+        Statement:
+          - Effect: Allow
+            Principal:
+              Service: lambda.amazonaws.com
+            Action:
+              - sts:AssumeRole
+      Policies:
+        - PolicyName: InlineSMTPLambdaExecutionRolePolicy
+          PolicyDocument:
+            Version: "2012-10-17"
+            Statement:
+                - Effect: Allow
+                  Action:
+                    - logs:CreateLogStream
+                    - logs:PutLogEvents
+                  Resource:
+                    - !Sub "arn:${AWS::Partition}:logs:${AWS::Region}:${AWS::AccountId}:log-group:/aws/lambda/SMTPCredentialsLambdaFunction"
+                    - !Sub "arn:${AWS::Partition}:logs:${AWS::Region}:${AWS::AccountId}:log-group:/aws/lambda/SMTPCredentialsLambdaFunction:log-stream:*"
+                - Effect: Allow
+                  Action:
+                    - ssm:PutParameter
+                    - ssm:DeleteParameter
+                  Resource:
+                    - !Sub "arn:${AWS::Partition}:ssm:${AWS::Region}:${AWS::AccountId}:parameter/smtp-username"
+                    - !Sub "arn:${AWS::Partition}:ssm:${AWS::Region}:${AWS::AccountId}:parameter/smtp-password"
+
+  SmtpKeyStack:
+    Type: AWS::CloudFormation::Stack
+    Properties:
+      TemplateURL: !Sub ${TemplateRootUrl}/aws-infra/v0.2.19/KMS/kms-key.yaml
+      TimeoutInMinutes: 5
+      Parameters:
+        AliasName: 'alias/smtp-credentials'
+        AdminPrincipalArns: !Join
+          - ','
+          - - !Sub 'arn:aws:iam::${AWS::AccountId}:root'
+            - !ImportValue
+              'Fn::Sub': '${AWS::Region}-workflows-nextflow-ci-service-account-ServiceUserArn'
+        UserPrincipalArns: !Join
+          - ','
+          - - !ImportValue
+              'Fn::Sub': '${AWS::Region}-workflows-nextflow-ci-service-account-ServiceUserArn'
+            - !GetAtt SmtpLambdaExecutionRole.Arn
+
+  SmtpLambdaLogGroup:
+    Type: AWS::Logs::LogGroup
+    Properties:
+      LogGroupName: !Join [ "/", [/aws, lambda, SMTPCredentialsLambdaFunction]]
+
+  SmtpLambdaFunction:
+    Type: AWS::Lambda::Function
+    DependsOn: SmtpLambdaLogGroup
+    Properties:
+      Description: Generates SMTP credentials and stores in Parameter Store
+      FunctionName: SMTPCredentialsLambdaFunction
+      Handler: index.lambda_handler
+      MemorySize: 128
+      Environment:
+        Variables:
+          KMS_KEY_ARN: !Sub 'arn:aws:kms:${AWS::Region}:${AWS::AccountId}:key/${SmtpKeyStack.Outputs.Key}'
```

Comment:
this is a dependency on `SmtpKeyStack`.  i would add `SmtpKeyStack` to `DependsOn` in this resource.

Reply:
Isn't that implicit as usual, since I'm using `!Sub`?

Reply:
I don't think `!Sub` makes it implicit, i think it depends on the property that you pass to.  these are arbitrary env vars so CFN may not enforce it as a dependency which may cause this to fail if timing is off.  my recommendation is to add to `DependsOn`

Reply:
I haven't seen this be a problem and from my reading of the docs this should work the same as any other use of implicit depends through Ref (which Sub is using) or GetAtt